File: internal/db/catalog_cache.go
package db

import (
	"context"
	_ "embed"
	"encoding/json"
	"os"
	"strconv"
	"strings"

	"reporting-db-migrations/internal/driver"
	"reporting-db-migrations/internal/fs"
)

//go:embed sql/catalog_cache_load.sql
var catalogCacheLoadSQL string

//go:embed sql/catalog_cache_invalidate.sql
var catalogCacheInvalidateSQL string

//go:embed sql/catalog_cache_save_meta.sql
var catalogCacheSaveMetaSQL string

//go:embed sql/catalog_cache_delete_all.sql
var catalogCacheDeleteAllSQL string

//go:embed sql/catalog_cache_insert_openjson.sql
var catalogCacheInsertOpenJSONSQL string

type catalogCacheJSONRow struct {
	K string `json:"k"`
	S string `json:"s"`
	G string `json:"g"`
	O string `json:"o"`
	P string `json:"p"`
}

// CatalogCacheEnabled reports whether persistent catalog cache is active (default on).
func CatalogCacheEnabled() bool {
	if strings.TrimSpace(os.Getenv("RMIG_CATALOG_CACHE")) == "0" {
		return false
	}
	if strings.TrimSpace(os.Getenv("RMIG_INSPECT_FULL")) == "1" {
		return false
	}
	return true
}

// SpotCheckCountFromEnv reads RMIG_CATALOG_SPOTCHECK (default 0).
func SpotCheckCountFromEnv() int {
	raw := strings.TrimSpace(os.Getenv("RMIG_CATALOG_SPOTCHECK"))
	if raw == "" {
		return 0
	}
	n, err := strconv.Atoi(raw)
	if err != nil || n < 0 {
		return 0
	}
	return n
}

// LayoutDigest is a stable SHA-256 hex digest of the layout scope (schemas, objects, transitions, checks).
func LayoutDigest(layout fs.Layout) string {
	return scopeKeySHA256Hex(scopeKey(layout))
}

func layoutObjectCount(layout fs.Layout) int {
	n := 0
	for _, obj := range layout.Objects {
		if obj != nil {
			n++
		}
	}
	return n
}

// tryLoadCatalogCache returns a full cached state when meta and row count match the layout.
func tryLoadCatalogCache(ctx context.Context, conn driver.Conn, layout fs.Layout, iscope InspectScope) (*State, bool, error) {
	if !CatalogCacheEnabled() {
		return nil, false, nil
	}
	want := layoutObjectCount(layout)
	digest := LayoutDigest(layout)
	rows, err := conn.QueryStringsContext(ctx, catalogCacheLoadSQL, []string{digest, strconv.Itoa(want)})
	if err != nil {
		// Tables may not exist yet (EnsureTables racing inspect); treat as cache miss.
		if isMissingCatalogTableErr(err) {
			return nil, false, nil
		}
		return nil, false, err
	}
	defer rows.Close()

	objects := make(map[string]Object, want)
	schemas := make(map[string]struct{})
	for rows.Next() {
		var key, schema, kind, object, parent string
		if err := rows.Scan(&key, &schema, &kind, &object, &parent); err != nil {
			return nil, false, err
		}
		objects[key] = Object{
			SchemaName: schema,
			Kind:       kind,
			ObjectName: object,
			ParentName: parent,
		}
		schemas[schema] = struct{}{}
	}
	if err := rows.Err(); err != nil {
		return nil, false, err
	}
	if len(objects) != want {
		return nil, false, nil
	}

	for k, obj := range iscope.StableObjects {
		objects[k] = obj
	}

	return &State{
		Schemas:      schemas,
		Objects:      objects,
		TableColumns: map[string][]TableColumn{},
	}, true, nil
}

func saveCatalogCache(ctx context.Context, conn driver.Conn, layout fs.Layout, state *State) error {
	if !CatalogCacheEnabled() || state == nil {
		return nil
	}
	want := layoutObjectCount(layout)
	if want == 0 {
		return nil
	}
	if len(state.Objects) < want {
		return nil
	}
	digest := LayoutDigest(layout)

	if _, err := conn.ExecContext(ctx, catalogCacheDeleteAllSQL); err != nil {
		if isMissingCatalogTableErr(err) {
			return nil
		}
		return err
	}
	payload, err := marshalCatalogCacheJSON(state.Objects, want)
	if err != nil {
		return err
	}
	if _, err := conn.ExecContext(ctx, catalogCacheInsertOpenJSONSQL, []string{payload, digest}); err != nil {
		return err
	}
	_, err = conn.ExecContext(ctx, catalogCacheSaveMetaSQL, []string{digest, strconv.Itoa(want)})
	return err
}

func invalidateCatalogCache(ctx context.Context, conn driver.Conn) {
	_, _ = conn.ExecContext(ctx, catalogCacheInvalidateSQL)
}

// InvalidateCatalogCacheForConn clears persisted catalog rows (e.g. after audit flush).
func InvalidateCatalogCacheForConn(ctx context.Context, conn driver.Conn) {
	invalidateCatalogCache(ctx, conn)
}

func marshalCatalogCacheJSON(objects map[string]Object, want int) (string, error) {
	rows := make([]catalogCacheJSONRow, 0, want)
	for k, obj := range objects {
		rows = append(rows, catalogCacheJSONRow{
			K: k,
			S: obj.SchemaName,
			G: obj.Kind,
			O: obj.ObjectName,
			P: obj.ParentName,
		})
	}
	if len(rows) != want {
		return "", nil
	}
	b, err := json.Marshal(rows)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

func isMissingCatalogTableErr(err error) bool {
	if err == nil {
		return false
	}
	msg := strings.ToLower(err.Error())
	return strings.Contains(msg, "catalog_cache") ||
		strings.Contains(msg, "catalog_meta") ||
		strings.Contains(msg, "invalid object name")
}

File: rust/crates/core/src/db/catalog_cache.rs
use crate::db::state::{catalog_object, CatalogState};
use crate::domain::ObjectKey;
use crate::driver::TimingConn;
use crate::error::{Error, Result};
use crate::sql;

pub async fn try_load(
    conn: &mut TimingConn,
    layout_digest: &[u8; 32],
    object_count: usize,
) -> Result<Option<CatalogState>> {
    if !cache_enabled() {
        return Ok(None);
    }
    let digest_hex = hex::encode(layout_digest);
    let count = object_count.to_string();
    let rows = match conn
        .query(
            sql::catalog::CACHE_LOAD,
            &[digest_hex.as_str(), count.as_str()],
        )
        .await
    {
        Ok(rows) => rows,
        Err(e) if missing_catalog_table(&e) => return Ok(None),
        Err(e) => return Err(e),
    };
    if rows.is_empty() {
        return Ok(None);
    }
    let mut state = CatalogState::default();
    for row in rows {
        merge_row(&mut state, &row)?;
    }
    if state.objects.len() != object_count {
        return Ok(None);
    }
    hydrate_schemas_from_objects(&mut state);
    crate::db::intern_catalog_state(&mut state);
    Ok(Some(state))
}

/// Load cached catalog when object count matches meta (git delta: digest may differ).
pub fn merge_load_rows(state: &mut CatalogState, rows: &[crate::driver::RowData]) -> Result<()> {
    for row in rows {
        merge_row(state, row)?;
    }
    hydrate_schemas_from_objects(state);
    Ok(())
}

fn hydrate_schemas_from_objects(state: &mut CatalogState) {
    for o in state.objects.values() {
        state.schemas.insert(o.schema.as_ref().to_lowercase());
    }
}

fn merge_row(state: &mut CatalogState, row: &crate::driver::RowData) -> Result<()> {
    let key = row.get_str(0).unwrap_or("");
    let schema = row.get_str(1).unwrap_or("");
    let kind = row.get_str(2).unwrap_or("");
    let name = row.get_str(3).unwrap_or("");
    let parent = row.get_str(4);
    if kind == "schema" {
        state.schemas.insert(schema.to_lowercase());
        return Ok(());
    }
    state.objects.insert(
        ObjectKey::from_normalized(key),
        catalog_object(schema, kind, name, parent),
    );
    Ok(())
}

pub async fn invalidate(conn: &mut TimingConn) -> Result<()> {
    if !cache_enabled() {
        return Ok(());
    }
    if let Err(e) = conn.exec(sql::catalog::CACHE_INVALIDATE).await {
        if missing_catalog_table(&e) {
            return Ok(());
        }
        return Err(e);
    }
    Ok(())
}

pub(crate) fn cache_enabled() -> bool {
    !matches!(
        std::env::var("RMIG_CATALOG_CACHE").as_deref(),
        Ok("0") | Ok("false")
    ) && std::env::var("RMIG_INSPECT_FULL").as_deref() != Ok("1")
}

pub(crate) fn missing_catalog_table(err: &Error) -> bool {
    let Error::Sql(msg) = err else {
        return false;
    };
    let m = msg.to_lowercase();
    m.contains("catalog_cache") || m.contains("catalog_meta") || m.contains("invalid object name")
}
